### ISSUE TITLE
How to train the SSD by 4 channel images

### ISSUE THREAD

**Issue by Jaekyumkim** [weiliu89/caffe#538]
Hi, I want to train the SSD by the 4 channel input images.
but when i try this, i can't change the weight size by 3 to 4.
so i got this error

I0417 17:24:51.615573 31632 layer_factory.hpp:77] Creating layer data
I0417 17:24:51.616021 31632 net.cpp:100] Creating Layer data
I0417 17:24:51.616040 31632 net.cpp:408] data -> data
I0417 17:24:51.616084 31632 net.cpp:408] data -> label
I0417 17:24:51.616952 31638 db_lmdb.cpp:35] Opened lmdb /media/user/4b3dfae8-c6b6-4a03-936d-d8fee7ff0b89/SSD/caffe/data/KITTI/depth_image/lmdb/KITTI_train_lmdb
I0417 17:24:51.648749 31632 annotated_data_layer.cpp:62] output data size: 32,4,150,600
I0417 17:24:51.718446 31632 net.cpp:150] Setting up data
I0417 17:24:51.718472 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718477 31632 net.cpp:157] Top shape: 1 1 1 8 (8)
I0417 17:24:51.718479 31632 net.cpp:165] Memory required for data: 46080032
I0417 17:24:51.718488 31632 layer_factory.hpp:77] Creating layer data_data_0_split
I0417 17:24:51.718503 31632 net.cpp:100] Creating Layer data_data_0_split
I0417 17:24:51.718509 31632 net.cpp:434] data_data_0_split <- data
I0417 17:24:51.718526 31632 net.cpp:408] data_data_0_split -> data_data_0_split_0
I0417 17:24:51.718538 31632 net.cpp:408] data_data_0_split -> data_data_0_split_1
I0417 17:24:51.718545 31632 net.cpp:408] data_data_0_split -> data_data_0_split_2
I0417 17:24:51.718552 31632 net.cpp:408] data_data_0_split -> data_data_0_split_3
I0417 17:24:51.718559 31632 net.cpp:408] data_data_0_split -> data_data_0_split_4
I0417 17:24:51.718564 31632 net.cpp:408] data_data_0_split -> data_data_0_split_5
I0417 17:24:51.718591 31632 net.cpp:408] data_data_0_split -> data_data_0_split_6
I0417 17:24:51.718672 31632 net.cpp:150] Setting up data_data_0_split
I0417 17:24:51.718680 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718684 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718689 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718693 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718698 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718703 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718708 31632 net.cpp:157] Top shape: 32 4 150 600 (11520000)
I0417 17:24:51.718713 31632 net.cpp:165] Memory required for data: 368640032
I0417 17:24:51.718732 31632 layer_factory.hpp:77] Creating layer conv1_1
I0417 17:24:51.718750 31632 net.cpp:100] Creating Layer conv1_1
I0417 17:24:51.718755 31632 net.cpp:434] conv1_1 <- data_data_0_split_0
I0417 17:24:51.718766 31632 net.cpp:408] conv1_1 -> conv1_1
F0417 17:24:51.773429 31639 annotated_data_layer.cpp:205] Check failed: std::equal(top_shape.begin() + 1, top_shape.begin() + 4, shape.begin() + 1) 
*** Check failure stack trace: ***
    @     0x7fd5fd7f35cd  google::LogMessage::Fail()
    @     0x7fd5fd7f5433  google::LogMessage::SendToLog()
    @     0x7fd5fd7f315b  google::LogMessage::Flush()
    @     0x7fd5fd7f5e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7fd5fdeb5190  caffe::AnnotatedDataLayer<>::load_batch()
    @     0x7fd5fde9e73f  caffe::BasePrefetchingDataLayer<>::InternalThreadEntry()
    @     0x7fd5fe0805a5  caffe::InternalThread::entry()
    @     0x7fd5f38f25d5  (unknown)
    @     0x7fd5ee5036ba  start_thread
    @     0x7fd5fc07082d  clone
    @              (nil)  (unknown)
./VGG_KITTI_SSD_600x150.sh: line 4: 31632 Aborted                 (core dumped) ./build/tools/caffe train --solver="/media/user/4b3dfae8-c6b6-4a03-936d-d8fee7ff0b89/SSD/caffe/data/KITTI/depth_image_code/SSD_4channel_structure/solver.prototxt" --gpu 0,1

and, when i try the force_color true, then the structure recognize the input just 3 channel images.
so, how can i train the SSD by 4 channel images.

**Comment by sujinpeng** [comment]
how do you calculate the mean_values?I also meet the problem

**Comment by hexiaoyi95** [comment]
You can try this: first get the mean value of  the three weight channels of the conv1_1  and repeat it four times to get the new conv1_1 with 4 channels. btw, you may want to copy the bias directly.

**Comment by sujinpeng** [comment]
I have an error:Check failed:std::equal(top_shape.begin()+1,top_shape.begin()+4,shape.begin()+1). how to solve it,please?

**Comment by Jaekyumkim** [comment]
@sujinpeng  i subtract the mean_value before making the lmdb dataset file.

**Comment by hexiaoyi95** [comment]
@sujinpeng if you are trying to do the same thing with @Jaekyumkim , then you need a new conv1_1 layer with bottom blob's shape is 4 channels

**Comment by Jaekyumkim** [comment]
@hexiaoyi95 but we can't make the 4 channel weight because there's no parameter to set the channel size. there's only num_output, pad, kernel_size, weight_filter, something like that.

**Comment by hexiaoyi95** [comment]
@Jaekyumkim you can define a new network with 4 channels input and make the name of conv1 differ from the ssd-model like "conv1_new".So you can finetune it based on the pre-train ssd caffemodel.The "conv1_new" layer will learn new params from your initialization and do not load it from the ssd-model .

**Comment by Jaekyumkim** [comment]
@hexiaoyi95  yes the conv1 the size of weight's channel is automatically initialized based on input channel size. But i didn't upload the pretrain weight based on VGGNet's. I just train the SSD based on initialization but the error come up like that. I think the conv1_1 weight size is not the problem.

**Comment by gabrieljcx** [comment]
Hey there,

I'm just doing the same thing as you but only with more channels right now.
I'm wondering if you've ever fixed your problem;)
If so, how did you fix the problem

Very Appreciated;)
